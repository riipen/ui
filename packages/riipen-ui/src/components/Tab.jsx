import clsx from "clsx";
import PropTypes from "prop-types";
import React from "react";

import ThemeContext from "../styles/ThemeContext";
import withClasses from "../utils/withClasses";

class Tab extends React.Component {
  static propTypes = {
    /**
     * If `true`, the tab indicator will be displayed.
     */
    active: PropTypes.bool,

    /**
     * An array of custom CSS classes to apply.
     */
    classes: PropTypes.array,

    /**
     * Determines the color of the indicator.
     */
    color: PropTypes.oneOf(["primary", "secondary"]),

    /**
     * If `true`, the tab will be disabled.
     */
    disabled: PropTypes.bool,

    /**
     * If `true`, will make the tab grow to use all the available space,
     */
    fullWidth: PropTypes.bool,

    /**
     * The icon element.
     */
    icon: PropTypes.elementType,

    /**
     * The label element.
     */
    label: PropTypes.node,

    /**
     * Callback fired when the tab is clicked.
     *
     * @param {object} event The event source of the callback
     * @param {any} value We default to the index of the child (number)
     */
    onClick: PropTypes.func,

    /**
     * The indicator orientation.
     */
    orientation: PropTypes.oneOf(["horizontal", "vertical"]),

    /**
     * The value provided when a click event is dispatched.
     */
    value: PropTypes.any.isRequired
  };

  static defaultProps = {
    active: false,
    color: "secondary",
    disabled: false,
    fullWidth: false,
    orientation: "horizontal"
  };

  static contextType = ThemeContext;

  handleClick = e => {
    const { onClick, value } = this.props;

    return onClick(e, value);
  };

  render() {
    const {
      active,
      classes,
      color,
      disabled,
      fullWidth,
      icon: Icon,
      label,
      orientation
    } = this.props;

    const theme = this.context;

    const className = clsx(
      "root",
      active ? `${color}-active` : null,
      active ? `${orientation}-active` : null,
      color,
      disabled ? "disabled" : null,
      fullWidth ? "full-width" : null,
      orientation,
      classes
    );

    return (
      <React.Fragment>
        <div className={className} onClick={this.handleClick}>
          {Icon && <Icon className={clsx("icon")} />}
          {label && <div className={clsx("label")}>{label}</div>}
        </div>
        <style jsx>{`
          .root {
            align-items: center;
            border-bottom: 3px solid transparent;
            cursor: pointer;
            display: inline-flex;
            justify-content: center;
            height: 100%;
            min-width: 72px;
            padding: ${theme.spacing(3)}px ${theme.spacing(6)}px;
            text-align: center;
            transition: all ${theme.transitions.duration.standard}ms;
          }

          .icon + * {
            margin-left: ${theme.spacing(2)}px;
          }

          .icon {
            margin-bottom: 1px;
          }

          .label {
            font-family: ${theme.typography.fontFamily};
            font-size: 14px;
            letter-spacing: 2px;
            line-height: 1.4;
            text-decoration: none;
            text-transform: uppercase;
            width: 100%;
          }

          .vertical {
            border-bottom: 0;
            border-right: 3px solid transparent;
            padding: ${theme.spacing()}px ${theme.spacing(5)}px;
            text-align: left;
          }

          .vertical-active {
            background-color: ${theme.palette.grey[50]};
          }

          .primary-active {
            border-color: ${theme.palette.primary.main};
            color: ${theme.palette.primary.main};
          }
          .primary:hover {
            color: ${theme.palette.primary.main};
          }
          .secondary-active {
            border-color: ${theme.palette.secondary.main};
            color: ${theme.palette.secondary.main};
          }
          .secondary:hover {
            color: ${theme.palette.secondary.main};
          }

          .disabled {
            opacity: 0.5;
            pointer-events: none;
          }

          .full-width {
            flex-basis: 0;
            flex-grow: 1;
            flex-shrink: 1;
            max-width: none;
          }
        `}</style>
      </React.Fragment>
    );
  }
}

export default withClasses()(Tab);
