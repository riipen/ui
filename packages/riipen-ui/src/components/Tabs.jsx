import clsx from "clsx";
import PropTypes from "prop-types";
import React from "react";

import withClasses from "../utils/withClasses";

class Tabs extends React.Component {
  static propTypes = {
    /**
     * The content of the component.
     */
    children: PropTypes.node,

    /**
     * An array of custom CSS classes to apply.
     */
    classes: PropTypes.array,

    /**
     * Determines the color of the indicator.
     */
    color: PropTypes.oneOf(["secondary", "primary"]),

    /**
     * The component used for the root node.
     * Either a string to use a DOM element or a component.
     */
    component: PropTypes.elementType,

    /**
     * Callback fired when the value changes.
     *
     * @param {object} event The event source of the callback
     * @param {any} value The value of the tab.
     */
    onChange: PropTypes.func,

    /**
     * The tabs orientation (layout flow direction).
     */
    orientation: PropTypes.oneOf(["horizontal", "vertical"]),

    /**
     * The value of the currently selected `Tab`.
     */
    value: PropTypes.any,

    /**
     *  Determines additional display behavior of the tabs:
     *
     *  - `standard` will render the default state.
     *  -`fullWidth` will make the tabs grow to use all the available space,
     *  which should be used for small views, like on mobile.
     */
    variant: PropTypes.oneOf(["standard", "fullWidth"])
  };

  static defaultProps = {
    color: "secondary",
    component: "div",
    orientation: "horizontal",
    variant: "standard"
  };

  handleChange = (e, value) => {
    const { onChange } = this.props;

    if (onChange) {
      onChange(e, value);
    }
  };

  render() {
    const {
      children,
      classes,
      color,
      component: Component,
      orientation,
      value,
      variant
    } = this.props;

    const className = clsx("root", orientation, classes);

    const childrenWithProps = React.Children.map(children, child => {
      if (!child) return null;

      return React.cloneElement(child, {
        active: child.props.value === value,
        color,
        fullWidth: variant === "fullWidth",
        onClick: this.handleChange,
        orientation
      });
    });

    return (
      <React.Fragment>
        <Component role="tablist" className={className}>
          <div className={clsx("scroller")}>
            <div className={clsx(orientation)}>{childrenWithProps}</div>
          </div>
        </Component>
        <style jsx>{`
          .root {
            display: flex;
            min-height: 48px;
            overflow: hidden;
          }

          .vertical {
            display: flex;
            flex-direction: column;
          }

          .scroller {
            display: inline-block;
            flex: 1 1 auto;
            overflow-x: auto;
            position: relative;
            white-space: nowrap;
          }
        `}</style>
      </React.Fragment>
    );
  }
}

export default withClasses(Tabs);
