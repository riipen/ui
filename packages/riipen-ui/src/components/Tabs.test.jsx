import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";

import Tab from "./Tab";
import Tabs from "./Tabs";

describe("<Tabs>", () => {
  it("renders without errors", () => {
    let error;

    try {
      mount(<Tabs />);
    } catch (theError) {
      error = theError;
    }

    expect(error).toEqual(undefined);
  });

  it("renders correct snapshot", () => {
    const wrapper = mount(<Tabs />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe("default props", () => {
    it("sets correct default props", () => {
      const defaultProps = new Tabs().type.defaultProps;

      const wrapper = mount(<Tabs />);

      const component = wrapper.find("Tabs");
      expect(component.props().breakpoint).toEqual(defaultProps.breakpoint);
      expect(component.props().color).toEqual(defaultProps.color);
      expect(component.props().component).toEqual(defaultProps.component);
      expect(component.props().orientation).toEqual(defaultProps.orientation);
      expect(component.props().variant).toEqual(defaultProps.variant);
    });
  });

  describe("breakpoint prop", () => {
    it("passes valid given breakpoint to children nodes", () => {
      const breakpoint = "xl";
      const child = <Tab label="Item one" value="one" />;

      const wrapper = mount(<Tabs breakpoint={breakpoint}>{child}</Tabs>);

      const childrenNode = wrapper.find("Tab");
      expect(childrenNode.props().breakpoint).toEqual(breakpoint);
    });

    it("gives an error with invalid breakpoint", () => {
      const errors = jest.spyOn(console, "error").mockImplementation();

      mount(<Tabs breakpoint="xs" />);

      expect(errors).toHaveBeenCalledTimes(1);
    });
  });

  describe("children prop", () => {
    it("renders all non null children", () => {
      const child1 = <Tab value="one" />;

      const wrapper = mount(
        <Tabs>
          {null}
          {child1}
        </Tabs>
      );

      const childrenNodes = wrapper.find("Tab");
      expect(childrenNodes.at(0).props().value).toEqual("one");
    });
  });

  describe("classes prop", () => {
    it("applies all classes to the root node", () => {
      const classes = ["classOne"];

      const wrapper = mount(<Tabs classes={classes} />);

      expect(
        wrapper
          .find("Tabs")
          .childAt(0)
          .hasClass(classes[0])
      ).toEqual(true);
    });

    it("appends higher order values to classes prop with withClass decorator", () => {
      const classes = ["riipen", "riipen-tabs"];

      const wrapper = mount(<Tabs />);

      expect(
        wrapper
          .find("Tabs")
          .props()
          .classes.sort()
      ).toEqual(classes.sort());
    });
  });

  describe("color prop", () => {
    it("passes valid given color to children", () => {
      const color = "primary";
      const child = <Tab label="Item one" value="one" />;

      const wrapper = mount(<Tabs color={color}>{child}</Tabs>);

      expect(wrapper.find("Tab").props().color).toEqual(color);
    });

    it("gives an error with invalid color", () => {
      const errors = jest.spyOn(console, "error").mockImplementation();

      mount(<Tabs color="yellow" />);

      expect(errors).toHaveBeenCalled();
    });
  });

  describe("component prop", () => {
    it("applies given component to the root node", () => {
      const component = "h3";

      const wrapper = mount(<Tabs component={component} />);

      expect(
        wrapper
          .find("Tabs")
          .childAt(0)
          .name()
      ).toEqual(component);
    });

    it("gives an error when given an invalid component", () => {
      const errors = jest.spyOn(console, "error").mockImplementation();
      const component = "abcd";

      mount(<Tabs component={component} />);

      expect(errors).toHaveBeenCalledTimes(1);
    });
  });

  describe("onChange prop", () => {
    it("sets onClick prop of children nodes", () => {
      const child = <Tab label="Item one" value="one" />;

      const wrapper = mount(<Tabs>{child}</Tabs>);

      expect(wrapper.find("Tab").props().onClick).not.toBeNull();
    });

    it("invokes handler when child Tab is clicked", () => {
      const onChange = jest.fn();
      const child = <Tab label="Item one" value="one" />;

      const wrapper = mount(<Tabs onChange={onChange}>{child}</Tabs>);

      wrapper
        .find("Tab")
        .childAt(0)
        .invoke("onClick")({ type: "click" });

      expect(onChange).toHaveBeenCalledWith(expect.any(Object), "one");
    });
  });

  describe("orientation prop", () => {
    it("passes given valid orientation to children", () => {
      const orientation = "vertical";
      const child = <Tab label="Item one" value="one" />;

      const wrapper = mount(<Tabs orientation={orientation}>{child}</Tabs>);

      expect(wrapper.find("Tab").props().orientation).toEqual(orientation);
    });

    it("sets valid custom orientation", () => {
      const orientation = "vertical";

      const wrapper = mount(<Tabs orientation={orientation} />);

      expect(
        wrapper
          .find("Tabs")
          .childAt(0)
          .childAt(0)
          .childAt(0)
          .hasClass(orientation)
      ).toEqual(true);
    });

    it("gives an error with invalid orientation", () => {
      const errors = jest.spyOn(console, "error").mockImplementation();

      mount(<Tabs orientation="forward" />);

      expect(errors).toHaveBeenCalled();
    });
  });

  describe("value prop", () => {
    it("sets the Tab with the corresponding value prop as active", () => {
      const child1 = <Tab label="Item one" value="one" />;
      const child2 = <Tab label="Item two" value="two" />;
      const value = "two";

      const wrapper = mount(
        <Tabs value={value}>
          {child1}
          {child2}
        </Tabs>
      );

      const childrenNodes = wrapper.find("Tab");
      expect(childrenNodes.at(0).props().active).toEqual(false);
      expect(childrenNodes.at(1).props().active).toEqual(true);
    });
  });

  describe("textTransform prop", () => {
    it("sets textTransform in children nodes", () => {
      const textTransform = "lowercase";
      const child = <Tab label="Item one" value="one" />;

      const wrapper = mount(<Tabs textTransform={textTransform}>{child}</Tabs>);

      expect(wrapper.find("Tab").props().textTransform).toEqual("lowercase");
    });

    it("gives an error with invalid textTransform", () => {
      const errors = jest.spyOn(console, "error").mockImplementation();

      mount(<Tabs textTransform="invalid" />);

      expect(errors).toHaveBeenCalled();
    });
  });

  describe("variant prop", () => {
    it("sets fullWidth to be true in children nodes when variant is fullWidth", () => {
      const variant = "fullWidth";
      const child = <Tab label="Item one" value="one" />;

      const wrapper = mount(<Tabs variant={variant}>{child}</Tabs>);

      expect(wrapper.find("Tab").props().fullWidth).toEqual(true);
    });

    it("sets fullWidth to be false in children nodes when variant is standard", () => {
      const variant = "standard";
      const child = <Tab label="Item one" value="one" />;

      const wrapper = mount(<Tabs variant={variant}>{child}</Tabs>);

      expect(wrapper.find("Tab").props().fullWidth).toEqual(false);
    });

    it("gives an error with invalid variant", () => {
      const errors = jest.spyOn(console, "error").mockImplementation();

      mount(<Tabs variant="invalid" />);

      expect(errors).toHaveBeenCalled();
    });
  });
});
