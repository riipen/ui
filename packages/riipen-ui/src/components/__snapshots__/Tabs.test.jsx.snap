// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tabs> renders correct snapshot 1`] = `
<withClasses(Tabs)>
  <Tabs
    breakpoint="sm"
    classes={
      Array [
        "riipen",
        "riipen-tabs",
      ]
    }
    color="secondary"
    component="div"
    orientation="horizontal"
    variant="standard"
  >
    <div
      className="jsx-3443918914 root horizontal riipen riipen-tabs"
      role="tablist"
    >
      <div
        className="jsx-3443918914 scroller"
      >
        <div
          className="jsx-3443918914 horizontal"
        />
      </div>
    </div>
    <JSXStyle
      id="3443918914"
    />
  </Tabs>
</withClasses(Tabs)>
`;
